pilot/pkg/networking/plugin/stats/stats.go

```diff
@@ -0,0 +1,143 @@
+	hcm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/wasm/v3"
+	protobuf "github.com/gogo/protobuf/types"
+	"github.com/golang/protobuf/ptypes"
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking"
+	"istio.io/istio/pilot/pkg/networking/plugin"
+	"istio.io/istio/pilot/pkg/networking/util"
+	xdsfilters "istio.io/istio/pilot/pkg/xds/filters"
+)
+
+// Plugin implements Istio Telemetry HTTP/TCP prometheus stats filter
+type Plugin struct{}
+
+// NewPlugin returns an instance of the HTTP/TCP prometheus stats plugin
+func NewPlugin() plugin.Plugin {
+	return Plugin{}
+}
+
+// OnInboundListener is called whenever a new HTTP/TCP prometheus stats filter is added to the Listener filter chain.
+func (p Plugin) OnInboundListener(in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	if in.Node.Type != model.SidecarProxy {
+		// Only care about sidecar.
+		return nil
+	}
+	return buildFilter("stats_inbound", in, mutable)
+}
+
+// OnInboundFilterChains is called whenever a plugin needs to setup the filter chains, including relevant filter chain
+// configuration, like FilterChainMatch and TLSContext.
+func (p Plugin) OnInboundFilterChains(in *plugin.InputParams) []networking.FilterChain {
+	return nil
+}
+
+// OnOutboundListener is called whenever a new HTTP/TCP prometheus stats filter is added to the Listener filter chain.
+func (p Plugin) OnOutboundListener(in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	return buildFilter("stats_outbound", in, mutable)
+}
+
+// OnInboundPassthrough is called whenever a new passthrough filter chain is added to the LDS output.
+// Can be used to add additional filters.
+func (p Plugin) OnInboundPassthrough(in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	return nil
+}
+
+// OnInboundPassthroughFilterChains is called whenever a plugin needs to setup custom pass through filter chain.
+func (p Plugin) OnInboundPassthroughFilterChains(in *plugin.InputParams) []networking.FilterChain {
+	return nil
+}
+
+// Build the HTTP and TCP prometheus stats telemetry filter based on the LisenterProtocol
+func buildFilter(vmId string, in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	for i := range mutable.FilterChains {
+		if in.ListenerProtocol == networking.ListenerProtocolHTTP || mutable.FilterChains[i].ListenerProtocol == networking.ListenerProtocolHTTP {
+			mutable.FilterChains[i].HTTP = append(mutable.FilterChains[i].HTTP, buildStatsHTTPFilter(vmId))
+		}
+		if in.ListenerProtocol == networking.ListenerProtocolTCP ||
+			mutable.FilterChains[i].ListenerProtocol == networking.ListenerProtocolTCP {
+			mutable.FilterChains[i].TCP = append(mutable.FilterChains[i].TCP, buildStatsTCPFilter(vmId))
+		}
+	}
+	return nil
+}
+
+func buildStatsHTTPFilter(vmId string) *hcm.HttpFilter {
```

Comment:
@howardjohn I put the buildFilter func in this stats.go file not in the xdsfilter filter.go file, since i need to pass "VmId" to build it based on inbound/outbound. This is not a "static" filter. I think probably for HTTPMx filter which the build func currently in xdsfilter folder, which is here: https://github.com/istio/istio/blob/master/pilot/pkg/xds/filters/filters.go#L171 we should put it in the metadataexchange.go file? And later on after all those filters configuration is clear, I'd like to extract those common stuff, probably we could make it as a single buildFilter func for all (this is not quite clear for me now, but the not static filter should put in their file or xdsfilter is an issue I'd like to know your idea)

Reply:
Aren't there only 2 VM ids? we can precompute one for each